sql/expression/function/locks.go

```diff
@@ -0,0 +1,223 @@
+	if nl.Child == nil {
+		return nil, nil
+	}
+
+	val, err := nl.Child.Eval(ctx, row)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if val == nil {
+		return nil, nil
+	}
+
+	lockName, ok := val.(string)
+
+	if !ok {
+		return nil, ErrIllegalLockNameArgType.New(nl.Child.Type().String(), nl.funcName)
+	}
+
+	return nl.logic(ctx, nl.ls, lockName)
+}
+
+// String implements the Stringer interface.
+func (nl *NamedLockFunction) String() string {
+	return fmt.Sprintf("%s(%s)", strings.ToUpper(nl.funcName), nl.Child.String())
+}
+
+// IsNullable implements the Expression interface.
+func (nl *NamedLockFunction) IsNullable() bool {
+	return nl.Child.IsNullable()
+}
+
+// WithChildren implements the Expression interface.
+func (nl *NamedLockFunction) WithChildren(children ...sql.Expression) (sql.Expression, error) {
+	if len(children) != 1 {
+		return nil, sql.ErrInvalidChildrenNumber.New(nl, len(children), 1)
+	}
+
+	return &NamedLockFunction{expression.UnaryExpression{Child:children[0]}, nl.ls, nl.funcName, nl.retType, nl.logic}, nil
+}
+
+// Type implements the Expression interface.
+func (nl *NamedLockFunction) Type() sql.Type {
+	return nl.retType
+}
+
+// ReleaseLockFunc is the function logic that is executed when the release_lock function is called.
+func ReleaseLockFunc(ctx *sql.Context, ls *sql.LockSubsystem, lockName string) (interface{}, error) {
+	err := ls.Unlock(ctx, lockName)
+
+	if err != nil {
+		if sql.ErrLockDoesNotExist.Is(err) {
+			return nil, nil
+		} else if sql.ErrLockNotOwned.Is(err) {
+			return int8(0), nil
+		}
+
+		return nil, err
+	}
+
+	return int8(1), nil
+}
+
+// IsFreeLockFunc is the function logic that is executed when the is_free_lock function is called.
+func IsFreeLockFunc(_ *sql.Context, ls *sql.LockSubsystem, lockName string) (interface{}, error) {
+	state, _ := ls.GetLockState(lockName)
+
+	switch state {
+	case sql.LockInUse:
+		return int8(0), nil
+	default: // return 1 if the lock is free.  If the lock doesn't exist yet it is free
+		return int8(1), nil
+	}
+}
+
+// IsUsedLockFunc is the function logic that is executed when the is_used_lock function is called.
+func IsUsedLockFunc(ctx *sql.Context, ls *sql.LockSubsystem, lockName string) (interface{}, error) {
+	state, owner := ls.GetLockState(lockName)
+
+	switch state {
+	case sql.LockInUse:
+		return owner, nil
+	default:
+		return nil, nil
+	}
+}
+
+// GetLock is a SQL function implementing get_lock
+type GetLock struct {
+	expression.BinaryExpression
+	ls *sql.LockSubsystem
+}
+
+// CreateNewGetLock returns a new GetLock object
+func CreateNewGetLock(ls *sql.LockSubsystem) func(e1, e2 sql.Expression) sql.Expression {
+	return func(e1, e2 sql.Expression) sql.Expression {
+		return &GetLock{expression.BinaryExpression{e1, e2}, ls}
+	}
+}
+
+// Eval implements the Expression interface.
+func (gl *GetLock) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	if gl.Left == nil {
+		return nil, nil
+	}
+
+	val0, err := gl.Left.Eval(ctx, row)
```

Comment:
leftVal and rightVal preferable to val0, val1

Reply:
changed